### ISSUE TITLE
Pretraining steps

### ISSUE THREAD

**Issue by Arij-Aladel** [google-research/longt5#7]
Hello !
could you please provide us with the detailed steps of generating the pretraining corpus ?

**Comment by Arij-Aladel** [comment]
I need to pretrain longt5 using another langauge , so providing pretraining steps will be appreciable. Thanks.

**Comment by stefan-it** [comment]
Hi @Arij-Aladel ,

I heavily think that pretraining steps are convenient with the T5/T5X workflow. That means you need to have a TensorFlow dataset, that must be uploaded to a GCP bucket. You can also use your own vocab. However, the complete end-to-end data preparation using an own corpus with own vocab is not really documented, so we did that for pretraining German T5 models:

Instructions for building an own TensorFlow dataset: https://github.com/GermanT5/corpus#tensorflow-dataset

Creating an own tokenizer using SPM: https://github.com/GermanT5/tokenizer

The crucial part is to write an own seqio task, because you need to use your previously trained SPM vocab in it: https://github.com/GermanT5/pre-training#custom-task

However, I'm not sure if this task can be used in longt5/Flaxformer library. And it would be awesome to know, how pretraining can be launched, because we're also interested in pretraining German LongT5 models.

**Comment by Arij-Aladel** [comment]
@stefan-it 
Hi, thanks for your reply. I think in your case for german you have mt5, is not enough?
in longt5 papaer it was mentioned that

> Regarding pre-training, we adopt the pretraining strategy in the PEGASUS (Zhang et al.,2019a) model.

in  (Zhang et al.,2019a) 
take a look into pretraining issues https://github.com/google-research/pegasus/issues?q=is%3Aissue+is%3Aopen+pretrain+

or pretraining steps on hugging face https://github.com/huggingface/transformers/issues/8536

So the steps are not well documented.
Beside  in the original PEGASUS paper another masking loss was added on top of the encoder to predict the <mask_2> . Is that also applicable to longT5? What are the steps?

**Comment by stefan-it** [comment]
We are pretraining from scratch (instead of using mT5 as init checkpoint), so we have an own vocab.

The PEGASUS parsing stuff is done here btw;

https://github.com/google-research/longt5/blob/aae210291ad21917a0a7cf33d42d87bd18e3db40/longt5/preprocessors.py#L139-L163

And should be modified when using an own vocab. 

Another missing thing in the documentation is the exact pretraining command 😅

**Comment by Arij-Aladel** [comment]
Yeah, you are right) @stefan-it thanks for the lightning! could you please share your email?

**Comment by stefan-it** [comment]
Hi @Arij-Aladel , feel free to write me `stefan@schweter.it` -> let's see if we can get pretraining running :hugs:

**Comment by stefan-it** [comment]
@mandyguo-xyguo Could you please help with the exact pretraining command :thinking:

**Comment by enpassanty** [comment]
I'm sure sure if this issue thread is till live.  I'm interested in pretraining from the published longT5 checkpoint (not form scratch) using my custom data set.  I assume I would need to merge SPM tokenizers (after creating a custom tokenizer for my data) before I start pre-training.  I am then going to finetune on my data that has a ling context.  Has anyone tried something similar and have any advise on a published recipe?
